ocsp.go

```diff
@@ -236,30 +276,75 @@ func signingParamsForPublicKey(pub interface{}, requestedSigAlgo x509.SignatureA
 			err = errors.New("x509: requested SignatureAlgorithm does not match private key type")
 			return
 		}
-		sigAlgo.Algorithm, hashFunc = details.oid, details.hash
-		if hashFunc == 0 {
+		if details.hash == crypto.Hash(0) {
 			err = errors.New("x509: cannot sign with hash function requested")
 			return
 		}
-		found = true
-		break
-	}
+		if details.isRSAPSS {
+			signerOpts = &rsa.PSSOptions{
+				SaltLength: rsa.PSSSaltLengthEqualsHash,
+				Hash:       details.hash,
+			}
+		} else {
+			signerOpts = details.hash
+		}
 
-	if !found {
-		err = errors.New("x509: unknown SignatureAlgorithm")
+		return signerOpts, pkix.AlgorithmIdentifier{
+			Algorithm:  details.oid,
+			Parameters: details.params,
+		}, nil
 	}
 
+	err = errors.New("x509: unknown SignatureAlgorithm")
 	return
 }
 
 // TODO(agl): this is taken from crypto/x509 and so should probably be exported
 // from crypto/x509 or crypto/x509/pkix.
-func getSignatureAlgorithmFromOID(oid asn1.ObjectIdentifier) x509.SignatureAlgorithm {
-	for _, details := range signatureAlgorithmDetails {
-		if oid.Equal(details.oid) {
-			return details.algo
+func getSignatureAlgorithmFromAI(ai pkix.AlgorithmIdentifier) x509.SignatureAlgorithm {
+	if !ai.Algorithm.Equal(oidSignatureRSAPSS) {
+		for _, details := range signatureAlgorithmDetails {
+			if ai.Algorithm.Equal(details.oid) {
+				return details.algo
+			}
 		}
+		return x509.UnknownSignatureAlgorithm
+	}
+
+	// RSA PSS is special because it encodes important parameters
+	// in the Parameters.
+	var params pssParameters
```

Comment:
The RSA-PSS part could go into its own function, and that path taken first instead of the negation just above. But not a blocker.

Reply:
The code above is almost a direct copy and paste from crypto/x509, even the comments.